### ISSUE TITLE
'Suite End' event not waiting for promises to finish

### ISSUE THREAD

**Issue by WillLuce** [mochajs/mocha#2368]
I'm writing a reporter to send results to a database and, because there is no `deferred` test emitter in the event that a suite bails out early, I'm collecting them and sending all of the deferred tests on `suite end`.

```
  runner.on('suite end', function (suite)
  {

    // Skip the global suite
    if (suite.title === '') { return; }
    /*
     * If the suite bailed out early and didn't run some of it's tests,
     * we want to make sure we track them as skipped.
     */
    if (remainingTests.length != 0)
    {
      console.log('       Suite ended with ' + remainingTests.length + ' tests remaining');

      remainingTests.forEach(function (skipped_test) {

        console.log(skipped_test.test_title);
        skipped_test.finish_time  = moment().format();
        skipped_test.duration     = 0;
        skipped_test.order_number = order_number;
        skipped_test.status       = 'skipped';

        order_number++;

        elastic.addResult(skipped_test)

      }); // end forEach

      remainingTests.length = 0;
    } // end if

  }); // end...suite end
```

The issue is that the `suite end` event doesn't wait for the `addResult` promises to resolve before exiting. This only becomes an problem if the final suite has deferred tests because the Node process exits too early. 

I have tried collecting the promises in an array and calling `Promises.all`, but no luck.

Currently, I'm solving this problem with a wait function on the `end` event, but this is hacky and unreliable. Am I missing something here? Shouldn't `suite end` be waiting for the promises to resolve?

**Comment by boneskull** [comment]
Unfortunately it isn't possible to do it this way, as you've found out.  Mocha will not *wait* for a reporter to finish with something, because it's simply emitting events; what happens thereafter is not its concern.  

`EventEmitter` listeners are executed synchronously; there's nothing you can do to tell an `EventEmitter` to wait for `Promise` resolution or a callback.  It's pub/sub, not RPC.

Your best bet may be to leverage a stream. [This module](https://www.npmjs.com/package/elasticsearch-streams) looks promising.

I think this is a use case we absolutely need to consider for the future.  I'm a little worried it will break the [js-reporters](https://github.com/js-reporters/js-reporters) spec, but maybe not.  When the time comes, someone should look into it further.

**Comment by WillLuce** [comment]
Thank you for getting back to me! I messed around with this for a few days before putting it up here. It's good to know that I wasn't doing anything stupid. I'll look into this feature.

**Comment by WillLuce** [comment]
In the meantime, it'd be nice to emit a `deferred` event for the leftover tests.

**Comment by boneskull** [comment]
@WillLuce Actually, I was poking around and found [this](https://github.com/mochajs/mocha/blob/master/lib/mocha.js#L494):

```js
  function done(failures) {
    if (reporter.done) {
      reporter.done(failures, fn);
    } else {
      fn && fn(failures);
    }
  }
```

Add a `done` method to your reporter and see what happens!

**Comment by WillLuce** [comment]
I think the issue happens when the suite bails out and runs [this](https://github.com/mochajs/mocha/blob/master/lib/runner.js#L739)

```
/**
 * Cleans up the references to all the deferred functions
 * (before/after/beforeEach/afterEach) and tests of a Suite.
 * These must be deleted otherwise a memory leak can happen,
 * as those functions may reference variables from closures,
 * thus those variables can never be garbage collected as long
 * as the deferred functions exist.
 *
 * @param {Suite} suite
 */
function cleanSuiteReferences(suite) {
  function cleanArrReferences(arr) {
    for (var i = 0; i < arr.length; i++) {
      delete arr[i].fn;
    }
  }

  if (isArray(suite._beforeAll)) {
    cleanArrReferences(suite._beforeAll);
  }

  if (isArray(suite._beforeEach)) {
    cleanArrReferences(suite._beforeEach);
  }

  if (isArray(suite._afterAll)) {
    cleanArrReferences(suite._afterAll);
  }

  if (isArray(suite._afterEach)) {
    cleanArrReferences(suite._afterEach);
  }

  for (var i = 0; i < suite.tests.length; i++) {
    delete suite.tests[i].fn;
  }
}
```

It removes all information about the tests. They aren't there anymore on `done`. If during this loop, they emitted a `deferred` event, we'd be in the clear.

**Comment by boneskull** [comment]
Couldn't you use `suite end` to save your data, then use `done()` to send it to ES?

**Comment by WillLuce** [comment]
Hmmm. I'll have to try that after I get my head wrapped around exactly where it should go. I have tried a few iterations of something like this:

```
    runner.on('suite', function (suite)
    {
        var waitForPromises = function (done)
        {
            Promise.all(promises)
                .then(function ()
                {
                    done();
                })
                .catch(function (e)
                {
                    console.error(e.message);
                    done(e);
                });
        };

        suite.afterEach(waitForPromises);
        suite.afterAll(waitForPromises);
    });
```

But, no dice. I also tried putting a variation of that in the `suite end` event. It's possible I'm not declaring that `done` fn correctly.

**Comment by WillLuce** [comment]
Where would I put that done for it to be called by this?[https://github.com/mochajs/mocha/blob/master/lib/mocha.js#L495](url)

```
  function done(failures) {
    if (reporter.done) {
      reporter.done(failures, fn);
    } else {
      fn && fn(failures);
    }
  }
```

I've tried declaring it outside the reporter function, then calling it inside, but again, the process exits before the function is executed...unless I put an explicit wait in.

**Comment by boneskull** [comment]
In your reporter you have a `runner` object; define `runner.done`:

```js
runner.done = function () {
  // ...
};
```

**Comment by WillLuce** [comment]
@boneskull I think this is the right approach, but I'm still running into a conceptual issue. From where is this called? I'm trying to get my reporter to wait for promises to be fulfilled. 

I've seen something like this before, but it hasn't worked for me
```
    runner.on('suite', function (suite)
    {
        var waitForPromises = function (done)
        {
            Promise.all(promises)
                .then(function ()
                {
                    done();
                })
                .catch(function (e)
                {
                    console.error(e.message);
                    done(e);
                });
        };

        suite.afterEach(waitForPromises);
        suite.afterAll(waitForPromises);
    });
```

**Comment by WillLuce** [comment]
@boneskull Interesting note on this. Somehow in the last week or so, skipped tests have started emitting `pending` events, which means I can subscribe to it and not issue so many promises at the same time immediately prior to exit. Oddly enough, this seems to bypass the issue being discussed here. 

Prior to this change, these tests did not emit any events, which was why I built my reporter the way I did. 

Did you guys change some things?
